src/blockchain/fork_choice.cpp

```diff
@@ -0,0 +1,444 @@
+                      .body = BlockBody{.attestations = attestations}};
+
+    // Apply state transition to get final post-state and compute state root
+    OUTCOME_TRY(stf_.processBlock(final_state, final_block));
+    final_block.state_root = sszHash(final_state);
+
+    // Store block and state in forkchoice store
+    auto block_hash = sszHash(final_block);
+    blocks_.emplace(block_hash, final_block);
+    states_.emplace(block_hash, std::move(final_state));
+
+    // update head (not in spec)
+    head_ = block_hash;
+
+    return final_block;
+  }
+
+
+  outcome::result<void> ForkChoiceStore::validateAttestation(
+      const SignedVote &signed_vote) {
+    auto &vote = signed_vote.data;
+
+    // Validate vote targets exist in store
+    if (not blocks_.contains(vote.source.root)) {
+      return Error::INVALID_ATTESTATION;
+    }
+    if (not blocks_.contains(vote.target.root)) {
+      return Error::INVALID_ATTESTATION;
+    }
+
+    // Validate slot relationships
+    auto &source_block = blocks_.at(vote.source.root);
+    auto &target_block = blocks_.at(vote.target.root);
+
+    if (source_block.slot > target_block.slot) {
+      return Error::INVALID_ATTESTATION;
+    }
+    if (vote.source.slot > vote.target.slot) {
+      return Error::INVALID_ATTESTATION;
+    }
+
+    // Validate checkpoint slots match block slots
+    if (source_block.slot != vote.source.slot) {
+      return Error::INVALID_ATTESTATION;
+    }
+    if (target_block.slot != vote.target.slot) {
+      return Error::INVALID_ATTESTATION;
+    }
+
+    // Validate attestation is not too far in the future
+    if (vote.slot > getCurrentSlot() + 1) {
+      return Error::INVALID_ATTESTATION;
+    }
+
+    return outcome::success();
+  }
+
+  outcome::result<void> ForkChoiceStore::processAttestation(
+      const SignedVote &signed_vote, bool is_from_block) {
+    signed_votes_[signed_vote.data.validator_id] = signed_vote;
+    // Validate attestation structure and constraints
+    BOOST_OUTCOME_TRY(validateAttestation(signed_vote));
+
+    auto &validator_id = signed_vote.data.validator_id;
+    auto &vote = signed_vote.data;
+
+    if (is_from_block) {
+      // update latest known votes if this is latest
+      auto latest_known_vote = latest_known_votes_.find(validator_id);
+      if (latest_known_vote == latest_known_votes_.end()
+          or latest_known_vote->second.slot < vote.slot) {
+        latest_known_votes_.insert_or_assign(validator_id, vote.target);
+      }
+
+      // clear from new votes if this is latest
+      auto latest_new_vote = latest_new_votes_.find(validator_id);
+      if (latest_new_vote != latest_new_votes_.end()
+          and latest_new_vote->second.slot <= vote.target.slot) {
+        latest_new_votes_.erase(latest_new_vote);
+      }
+    } else {
+      // forkchoice should be correctly ticked to current time before importing
+      // gossiped attestations
+      if (vote.slot > getCurrentSlot() + 1) {
```

Comment:
`getCurrentSlot()` should return slot from `time_`, not `clock.now()`
```suggestion
      if (vote.slot > getCurrentSlot()) {
```